ADC: STM32: Add device tree config for clock and prescaler

FRASTM opened the pull request:
This PR adds a ADC clock plus a prescaler to the ADC of the stm32 mcus.

With a `st,adc-prescaler` property in the ADC node, it is now possible to select the exact ADC clock scheme to be SYNC or ASYNC and its divider. By combining both information into one property, as follows : 
          `st,adc-prescaler = <ASYNC 16>;`
the clock prescaler is matching the LL_ADC_CLOCK_ASYNC_DIV16

Note that the **#include <zephyr/dt-bindings/adc/stm32_adc.h>** is required when setting the `st,adc-prescaler`.

It depends on the stm32 serie (refer to the RefMan) and if property is not set, a default config is applied by the LL_ADC_SetClock or LL_ADC_SetCommonClock at ADC driver init

It is increasing and replacing the https://github.com/zephyrproject-rtos/zephyr/pull/44130


Fixes https://github.com/zephyrproject-rtos/zephyr/issues/40955


Signed-off-by: Francois Ramu [francois.ramu@st.com](mailto:francois.ramu@st.com)

Commit:
dts: bindings: adc stm32 has a prescaler entry for the ADC clock

This is a new parameter to give the type of ADC clock source
asynchronous or synchronous.
And a new parameter to divide the ADC clock by the ADC_PRESC.
Both properties are combined to give the exact the ADC clock:
LL_ADC_CLOCK_ASYNC_DIVx or LL_ADC_CLOCK_SYNC_PCLK_DIVx
depending on the serie (not all are allowed).

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

FRASTM committed:
include: bindings: adc for stm32 macro

Values to help defining the Device tree macro
when generating the ADC clock prescaler scheme.
Macro to define if clock is common from the previous value.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
tests: drivers: adc driver testing on stm32 various boards

This commit shows an example for defining a particular
ADC clock scheme and prescaler of a stm32 board .
Define an overlay to address the adc device tree.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented on dts/bindings/adc/st,stm32-adc.yaml:
```diff
@@ -21,6 +21,21 @@ properties:
     "#io-channel-cells":
       const: 1
 
+    st,prescaler:
```
Let's be a bit more specific for this property so users don't mix with existing STM32 timer's property 'st,prescaler'
```suggestion
    st,adc-prescaler:
```

FRASTM replied:
done

erwango commented on dts/bindings/adc/st,stm32-adc.yaml:
```diff
@@ -21,6 +21,21 @@ properties:
+      type: array
+      description: |
+        The ADC Clock specifier at the input of the peripheral.
+        Depending on the soc serie, it is ASYNChronous specific clock
+        or SYNChronous derived from the AHB or APB.
```
```suggestion
        On some series, it could be either ASYNChronous (derived from the domain clock)
        or SYNChronous (derived from the AHB or APB).
        In each case, a specific prescaler can be applied to reduce the input frequency.
```

FRASTM replied:
changed

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -1174,6 +1176,34 @@ static const struct adc_driver_api api_stm32_driver_api = {
+#define STM32_ADC_CLOCK_IS_SYNC(x, prop) IS_EQ(STM32_ADC_CLOCK(x, prop), SYNC)
+
+/* Macro to set the prefix depending on the 1st element */
+#define STM32_ADC_DO_PREFIX(x, prop)				\
+	COND_CODE_1(IS_EQ(STM32_ADC_CLOCK(x, prop), SYNC),	\
+		prescaler_sync,					\
+		prescaler_async)
```
```suggestion
#define STM32_ADC_PREFIX(x, prop)				\
	COND_CODE_1(IS_EQ(STM32_ADC_CLOCK(x, prop), SYNC),	\
		LL_ADC_CLOCK_SYNC_PCLK_DIV,					\
		LL_ADC_CLOCK_ASYNC_DIV)
```

FRASTM replied:
Using directly LL_ADC_xxx prefix to simplify and change macro name

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -1174,6 +1176,34 @@ static const struct adc_driver_api api_stm32_driver_api = {
+ * The st,prescaler property must declare 2 elements
+ * else choosing the default prescaler clock and div by 4
+ */
+#define DT_INST_PRESC_PROP(x, prop)					\
```
I think it is preferred to use _STM32_ namespace as DT_INST_XXX_PROP might be reserved for Zephry generic macros.
```suggestion
#define DT_STM32_PRESC_PROP(x, prop)					\
```

FRASTM replied:
ok, changing macro name

erwango replied:
Also, 'prop' argument should not be required (always the same prop, and is matching the macro name `PRESC` )

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -1174,6 +1176,34 @@ static const struct adc_driver_api api_stm32_driver_api = {
+ * else choosing the default prescaler clock and div by 4
+ */
+#define DT_INST_PRESC_PROP(x, prop)					\
+	COND_CODE_1(IS_EQ(DT_INST_PROP_LEN(x, prop), 2),		\
```
What about ?
```suggestion
	COND_CODE_1(DT_INST_NODE_HAS_PROP(x, st_adc_prescaler),		\
```

FRASTM replied:
COND_CODE_1(CONFIG_FLAG, ...) accepts CONFIG_FLAG to be IS_EQ(something) and  IS_EQ(property len, 2) is a valid CONFIG_FLAG
--> DT_INST_NODE_HAS_PROP(x, st_adc_prescaler) is not a _CONFIG_FLAG_ for COND_CODE_1

FRASTM replied:
next step: add a check on st,adc-prescaler len and 1st element to be SYNC or ASYNC if present

erwango replied:
Weird, there are >50 occurences of `COND_CODE_1(DT_INST_NODE_HAS_PROP(` in the tree.

FRASTM replied:
ok with COND_CODE_1(DT_INST_NODE_HAS_PROP()

FRASTM commented:
next step: add a check on st,adc-prescaler len and 1st element to be SYNC or ASYNC if present

erwango commented on tests/drivers/adc/adc_api/boards/b_u585i_iot02a.overlay:
```diff
@@ -0,0 +1,10 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/dt-bindings/adc/stm32_adc.h>
```
Let's move this to `dts/arm/st/x*/stm32x*.dtsi` files

FRASTM replied:
moving the zephyr/dt-bindings/adc/stm32_adc.h from the test overlay file to the stm32x.dtsi for stm32u5x and stm32l5 series.

erwango replied:
Why not extending to all other STM32 series ?

FRASTM replied:
added to all the stm32 soc

FRASTM commented:
Add a commit (ceb218b5144fb80cf32c949404052615d59a0d9c)  to define a macro for check the correct format of the `st,adc-prescaler` property.
_COND_CODE_1 : To prevent the preprocessor from treating commas as argument separators, the _if_1_code and _else_code expressions must be inside brackets/parentheses: (). These are stripped away during macro expansion._

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -1174,6 +1176,60 @@ static const struct adc_driver_api api_stm32_driver_api = {
+			(BUILD_ASSERT((					\
+				((STM32_ADC_CLOCK(x, prop) == SYNC) ||	\
+				(STM32_ADC_CLOCK(x, prop) == ASYNC)) &&	\
+				((STM32_ADC_DIV(x, prop) >= 1) &&	\
+				(STM32_ADC_DIV(x, prop) <= 256))),	\
```
I propose to drop check on value.
This check is quite rough (in a lot of case a value passing this check could be wrong) and 
user will anyway get a compilation error if trying to use any value not matching `LL_ADC_CLOCK_ASYNC_DIVX` defined by Cube (`LL_ADC_CLOCK_ASYNC_DIV777` for instance).

FRASTM replied:
ok removed

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -1174,6 +1176,60 @@ static const struct adc_driver_api api_stm32_driver_api = {
+ */
+#define ADC_STM32_HAS_PROP(x, prop)					\
+	IF_ENABLED(DT_INST_NODE_HAS_PROP(x, prop),			\
+		(COND_CODE_1(						\
+			IS_EQ(DT_INST_PROP_LEN(x, prop), 2),		\
+			(BUILD_ASSERT((					\
+				((STM32_ADC_CLOCK(x, prop) == SYNC) ||	\
+				(STM32_ADC_CLOCK(x, prop) == ASYNC)) &&	\
+				((STM32_ADC_DIV(x, prop) >= 1) &&	\
+				(STM32_ADC_DIV(x, prop) <= 256))),	\
+				"Invalid ADC "#prop" property")),	\
+			(BUILD_ASSERT((DT_INST_PROP_LEN(x, prop) == 2),	\
+				"Invalid ADC "#prop" property length")))))
```
Since there is no real reason to use COND_CODE logic here I wonder if this could be simplified with the following :
```suggestion
IF_ENABLED(DT_INST_NODE_HAS_PROP(x, prop), \
			(BUILD_ASSERT(FOO) || (BUILD_ASSERT(BAR))
```

FRASTM replied:
ORing BUILD_ASSERT in that way, will give a compilation error in on or the other case:
```
error: expected ';' before '||' token
 1218 |    || BUILD_ASSERT 
```

FRASTM commented:
without check the values of the divider in the st,adc-prescaler property

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -1174,6 +1176,58 @@ static const struct adc_driver_api api_stm32_driver_api = {
+ * 1st one is SYNC or ASYNC (not all series have both)
+ * 2nd one for the divider (not all series have all values)
+ */
+#define ADC_STM32_HAS_PROP(x, prop)					\
+	IF_ENABLED(DT_INST_NODE_HAS_PROP(x, prop),			\
+		(COND_CODE_1(						\
+			IS_EQ(DT_INST_PROP_LEN(x, prop), 2),		\
+			(BUILD_ASSERT((					\
+				(STM32_ADC_CLOCK(x, prop) == SYNC) ||	\
+				(STM32_ADC_CLOCK(x, prop) == ASYNC)),	\
+				"Invalid ADC "#prop" property")),	\
+			(BUILD_ASSERT((DT_INST_PROP_LEN(x, prop) == 2),	\
+				"Invalid ADC "#prop" property length")))))
```
Ok, so following should work.
```
#define ADC_STM32_HAS_PROP(x, prop)					\
	IF_ENABLED(DT_INST_NODE_HAS_PROP(x, prop),			\
			(BUILD_ASSERT(					\
				(STM32_ADC_CLOCK(x, prop) == SYNC) ||	\
				(STM32_ADC_CLOCK(x, prop) == ASYNC),	\
				"Invalid ADC "#prop" property");	\
			BUILD_ASSERT((DT_INST_PROP_LEN(x, prop) == 2),	\
				"Invalid ADC "#prop" property length")));
```

erwango replied:
Also, macro is quite specific to st,adc-prescaler binding.
Naming should be adapted to make it more explicit:
`DT_STM32_CHECK_PRESC` or similar.

FRASTM replied:
changing to `ADC_STM32_CHECK_PRESC`

FRASTM replied:
change macro to remove the COND_CODE_1. 
Note that it must  first check the length then the 1st element (SYNC/ASYNC) else the 
`st,adc-prescaler = <3>;`  is not detected as a length error

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -1174,6 +1176,58 @@ static const struct adc_driver_api api_stm32_driver_api = {
+ * If the property is not defined, the default prescaler with DIV 4 is used.
+ */
+#if defined(CONFIG_SOC_SERIES_STM32F1X)
+#define DT_STM32_PRESC_PROP(x, prop) 0
+#else
+#define DT_STM32_PRESC_PROP(x, prop)					\
+	COND_CODE_1(DT_INST_NODE_HAS_PROP(x, prop),			\
+	_CONCAT(STM32_ADC_PREFIX(x, prop), STM32_ADC_DIV(x, prop)),	\
+	_CONCAT(prescaler_default, 4))
```
All these macros are specific to this prescaler property.
I think we should remove the 'prop' argument, since no other prop could be used as argument.

FRASTM replied:
changing macro name to `DT_STM32_PRESC`

FRASTM commented:
add the `#include <zephyr/dt-bindings/adc/stm32_adc.h>` to the dstsi for all the stm32 series
rename and reformat macro

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -1174,6 +1176,56 @@ static const struct adc_driver_api api_stm32_driver_api = {
+#define STM32_ADC_DIV(x, prop)		DT_INST_PROP_BY_IDX(x, prop, 1)
+
+/*
+ * Macro to check the property is valid with two elements:
+ * 1st one is SYNC or ASYNC (not all series have both)
+ * 2nd one for the divider (not all series have all values)
```
To be updated

FRASTM replied:
comment changed to
```
 * 1st one is SYNC or ASYNC
 * 2nd one for the divider
```

FRASTM commented:
remove 'prop' parameter from macro and access the `st_adc_prescaler` directly

galak commented on dts/bindings/adc/st,stm32-adc.yaml:
```diff
@@ -21,6 +21,22 @@ properties:
+        - first:  SYNC or ASYNC to select synchronous or asynchronous specific clock
+        - second: divider which is an integer from 1 to 256
+        (some stm32 may have only limited nb of divider. For example on stm32F2 :
+        SYNC of 4 (possibly 2, 4, 6, 8) will give st,prescaler = <SYNC 4>;
```
It feels weird to mix data types like this in an array.  Feels like there should be:
`st,adc-prescale-type' (SYNC/ASYNC) and `st-adc-prescaler`.

erwango replied:
Is that blocking? 
FWIW, this is what we're doing already for pinctrl:
`pinmux = <STM32_PINMUX('A', 3, ANALOG)>;`

We'll need to add much more props in these ADC bindings in the end, so if there's a way to save a few, I would be pleased. Anyway, up to you!

FRASTM replied:
Giving two properties (st,adc-prescaler-type' and 'st,adc-prescaler') was a (much more simpler) option that was not retained to combine both information in a single property.

erwango replied:
@galak Is that bocking ?

erwango replied:
@galak ping. Also note that one of the reason for combining these props is that both registers are strictly linked:
`he ADC prescaler value is applied only when CKMODE[1:0] = 0b00.` and they combine in a single parameter of the function `LL_ADC_SetCommonClock`.
So it's possible to split in 2 properties, but I do think current proposal makes things easier for the user.

erwango replied:
@mbolivar-nordic thoughts ?

mbolivar-nordic replied:
> @mbolivar-nordic thoughts ?

No thoughts from me -- I'm not maintaining the individual bindings anymore, so you'll have to get a reply from @galak. 72637ac746a7fe92bc2d76f21e8c4a53eceb17bd

FRASTM commented:
fix doc error in include/zephyr/dt-bindings/adc/stm32_adc.h

mbolivar-nordic commented on drivers/adc/adc_stm32.c:
```diff
@@ -1174,6 +1176,56 @@ static const struct adc_driver_api api_stm32_driver_api = {
+ * If the property is not defined, the default prescaler with DIV 4 is used.
+ */
+#if defined(CONFIG_SOC_SERIES_STM32F1X)
+#define DT_STM32_PRESC(x) 0
```
The `DT_` namespace is reserved for devicetree.h. I would prefer if you renamed it to STM32_DT_PRESC.

FRASTM replied:
changed toSTM32_ADC_DT_PRESC

FRASTM commented:
Because the prescaler is now selectable (SYNC_x or ASYNC_x), some other conditions apply like for example on stm32WL : _"ADC common clock asynchronous prescaler is applied to each ADC instance if the corresponding ADC instance clock is set to clock source asynchronous."_

FRASTM committed:
dts: arm: st dts for stm32 series include stm32_adc.h bindings

When using the st,adc-prescaler, property in a DTS,
the stm32 device tree must include the
zephyr/dt-bindings/adc/stm32_adc.h to get the
SYNC and ASYNC values.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: adc: stm32 driver checks the st,adc-prescaler and type format

This commit adds a macro to control the format of the
st,adc-prescaler and type properties if present in the DTS.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
new version of the adc stm32 driver and  DTS where the adc clock prescaler has two properties to be combined:

- st,adc-prescaler-type:  <SYNC> or <ASYNC>
- st,adc-prescaler:  value 1 .. 256

DTS now looks like :
```
&adc1 {
	st,adc-prescaler-type = <SYNC>;
	st,adc-prescaler = <64>;
};
```
(not all combination valid though)

FRASTM committed:
drivers: adc: stm32 build the LL_ADC_CLOCK from the DTS property

Get the adc-prescaler and  adc-prescaler-type properties
in the DTS to be the ADC clock asynchronous or synchronous
and combine with the divider to have the exact clock scheme
LL_ADC_CLOCK_ASYNC_DIVx or LL_ADC_CLOCK_SYNC_PCLK_DIVx.
This depends on the stm32 serie (stm32F1 has no prescaler capability).

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -262,6 +263,7 @@ struct adc_stm32_cfg {
 	void (*irq_cfg_func)(void);
 	struct stm32_pclken pclken;
 	const struct pinctrl_dev_config *pcfg;
+	uint32_t clock_prescaler;
```
```suggestion
	const uint32_t clock_prescaler;
```

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -1289,6 +1290,41 @@ static const struct adc_driver_api api_stm32_driver_api = {
 	.ref_internal = STM32_ADC_VREF_MV, /* VREF is usually connected to VDD */
 };
 
+/*
+ * Prefix default when not configured by the DTS
+ * depending on the serie (stm32F1 has none)
+ */
+#if defined(CONFIG_SOC_SERIES_STM32L1X) || \
+	defined(CONFIG_SOC_SERIES_STM32U5X)
+#define prescaler_default	LL_ADC_CLOCK_ASYNC_DIV
+#elif !defined(CONFIG_SOC_SERIES_STM32F1X)
+#define prescaler_default	LL_ADC_CLOCK_SYNC_PCLK_DIV
+#endif
```
It seems to me it would be easier using default in bindings.
This is an allowed case for default usage "defaults which match the vendor-specified power-on reset value (as long as they are independent from other properties)"
